protocol/src/main/java/org/mifos/chatbot/protocol/facebookMessenger/FacebookMessengerChatService.java

```diff
@@ -0,0 +1,275 @@
+        }
+    }
+
+    @RequestMapping(value = {"/", ""}, method = RequestMethod.POST)
+    public ResponseEntity<Void> handleCallback(@RequestBody final String payload, @RequestHeader(SIGNATURE_HEADER_NAME) final String signature) {
+        try {
+            this.messenger.onReceiveEvents(payload, of(signature), event -> {
+                if (event.isTextMessageEvent()) {
+                    handleTextMessageEvent(event.asTextMessageEvent());
+                } else if (event.isAttachmentMessageEvent()) {
+                    handleAttachmentMessageEvent(event.asAttachmentMessageEvent());
+                } else if (event.isQuickReplyMessageEvent()) {
+                    handleQuickReplyMessageEvent(event.asQuickReplyMessageEvent());
+                } else if (event.isPostbackEvent()) {
+                    handlePostbackEvent(event.asPostbackEvent());
+                } else if (event.isAccountLinkingEvent()) {
+                    handleAccountLinkingEvent(event.asAccountLinkingEvent());
+                } else if (event.isOptInEvent()) {
+                    handleOptInEvent(event.asOptInEvent());
+                } else if (event.isMessageEchoEvent()) {
+                    handleMessageEchoEvent(event.asMessageEchoEvent());
+                } else if (event.isMessageDeliveredEvent()) {
+                    handleMessageDeliveredEvent(event.asMessageDeliveredEvent());
+                } else if (event.isMessageReadEvent()) {
+                    handleMessageReadEvent(event.asMessageReadEvent());
+                } else {
+                    handleFallbackEvent(event);
+                }
+            });
+            return ResponseEntity.status(HttpStatus.OK).build();
+        } catch (MessengerVerificationException e) {
+            log.error(e.getStackTrace().toString());
+            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+        } catch (NotImplementedException e) {
+            log.error(e.getStackTrace().toString());
+            return ResponseEntity.status(HttpStatus.OK).build();
+        }
+    }
+
+    private boolean authUser(String username, String password) {
+        HttpClient client = HttpClientBuilder.create().build();
+        HttpPost post = new HttpPost(settings.getApiUrl() + "/authentication?username=" + username + "&password=" + password + "&tenantIdentifier=mifos");
+        try {
+            HttpResponse response = client.execute(post);
+            if(response.getStatusLine().getStatusCode() == 200) {
+                return true;
+            }
+        } catch (Exception e) {
+            log.error(e.toString());
+        }
+        return false;
+    }
+
+    private static String base64Encode(String input) {
+        final byte[] authBytes = input.getBytes(StandardCharsets.UTF_8);
+        return Base64.getEncoder().encodeToString(authBytes);
+    }
+
+    private void handleTextMessageEvent(TextMessageEvent event) {
+        final String messageText = event.text();
```

Comment:
Can we please:
1. modularize this entire function into smaller logical tasks
2. document the modularization for a clearer understanding

Reply:
this method is now in 3 parts, to handle login, logout and generate response.

Reply:
Looks cleaner, this is good